### PR TITLE
✨ Changing Inline Page Attachment Button UI 

### LINKED ISSUE #32770: [Page attachments] [Inline] Image preview UI
Accept one or two images.
Text is optional.

Screen Shot 2021-02-19 at 5 18 45 PM
Screen Shot 2021-02-19 at 5 20 47 PM

@ampproject/wg-stories

### PR DESCRIPTION
**Pull request by raxsha**
- Separates outlink & inline code paths.
- Adds new chip.
- Added customizable image options: 0, 1, or 2 custom images.
- Added light theme arrow.
- Protected changes with flag.

Fixes #32770. 

Firebase demo: https://amp-story-inline-cta.web.app/examples/amp-story/attachment.html
To see it, be sure to set the feature flag in the console with: AMP.toggleExperiment('amp-story-page-attachment-ui-v2');.
Feature is on page 1.

Preview:
inline-cta-demo

One image:
Screen Shot 2021-03-09 at 2 30 44 PM

No text:
Screen Shot 2021-03-09 at 2 31 17 PM

### CODE REVIEW

**Comment by newmuis** [comment]
What's the API for setting the images?  Maybe I'm overlooking it in the source.

I think having multiple attributes (e.g. cta-image and cta-image2) would actually just be simplest, since we can check it easily from validation

**Comment by raxsha** [comment]
> What's the API for setting the images? Maybe I'm overlooking it in the source.
> 
> I think having multiple attributes (e.g. cta-image and cta-image2) would actually just be simplest, since we can check it easily from validation

Currently I have the `cta-img` attribute expecting a comma-separated list. If the attribute is not provided, we default to the old CTA UI. If there's 1 image provided, we add 1 image, if there's more than 1, we add the first and the second image provided and ignore the rest.

File: examples/amp-story/attachment.html [hunk 1]
```diff
@@ -51,7 +51,14 @@
             <span data-text-background-color="#00A3DB">Relentlessly pursues moth annoy owner until he gives you food say meow repeatedly until belly rubs, feels good drool that box?</span>
           </p>
         </amp-story-grid-layer>
-        <amp-story-page-attachment data-title="Lorem ipsum" layout="nodisplay">
+        <amp-story-page-attachment
+          data-title="Lorem ipsum"
+          layout="nodisplay"
+          cta-text="Call To Action"
+          cta-img="/extensions/amp-story/img/sweater.jpg"
+          cta-img-2="/extensions/amp-story/img/boot.jpg"
```
Comment (processprocess) [inline examples/amp-story/attachment.html:59]: a11y:
Publishers need a way to specify alt text for these images.
Each image needs a `-alt` attribute, like `cta-image-alt` and `cta-img-2-alt`.

File: extensions/amp-story/1.0/amp-story-open-page-attachment.js [hunk 2]
```diff
@@ -38,13 +40,50 @@ export const buildOpenAttachmentElement = (element) =>
...
+ * @return {!Element}
+ */
+export const buildOpenInlineAttachmentElement = (element) =>
+  htmlFor(element)`
+    <a class="
+        i-amphtml-story-page-open-attachment i-amphtml-story-system-reset"
+        role="button">
+      <div class="i-amphtml-story-inline-page-attachment-chip">
+        <div class="i-amphtml-story-inline-page-attachment-img"></div>
+        <div class="i-amphtml-story-inline-page-attachment-img-2"></div>
```
Comment (processprocess) [inline extensions/amp-story/1.0/amp-story-open-page-attachment.js:53]: a11y:
These need to be visible to screen readers.
This can be achieved by adding `role="image"`
They also an `alt` attribute. `alt` can be empty by default and `cta-image-alt` or `cta-img-2-alt` can be propagated to it.

File: extensions/amp-story/1.0/amp-story-open-page-attachment.css [hunk 3]
```diff
@@ -166,3 +166,61 @@
   text-shadow: 0px 0px 6px rgba(0, 0, 0, 0.36) !important;
   white-space: nowrap !important;
 }
+
+/** Inline attachment specifications */
```
Comment (processprocess) [inline extensions/amp-story/1.0/amp-story-open-page-attachment.css:170]: This comment can be more direct, like 
`For amp-story-inline-page-attachment-v2 experiment.`

File: extensions/amp-story/1.0/amp-story-open-page-attachment.css [hunk 4]
```diff
@@ -166,3 +166,61 @@
...
+.i-amphtml-story-inline-page-attachment-chip {
+  display: flex !important;
+  position: relative !important;
```
Comment (processprocess) [inline extensions/amp-story/1.0/amp-story-open-page-attachment.css:173]: Might not need this.

File: extensions/amp-story/1.0/amp-story-open-page-attachment.css [hunk 5]
```diff
@@ -166,3 +166,61 @@
   text-shadow: 0px 0px 6px rgba(0, 0, 0, 0.36) !important;
   white-space: nowrap !important;
 }
+
+/** Inline attachment specifications */
+.i-amphtml-story-inline-page-attachment-chip {
+  display: flex !important;
```
Comment (processprocess) [inline extensions/amp-story/1.0/amp-story-open-page-attachment.css:172]: Might not need this since we're using margin to center `i-amphtml-story-inline-page-attachment-arrow`.

**Comment by raxsha** [comment]
UI LGTM from Hong ✅

File: extensions/amp-story/1.0/amp-story-open-page-attachment.css [hunk 6]
```diff
@@ -166,3 +166,61 @@
...
+.i-amphtml-story-inline-page-attachment-arrow {
+  height: 32px !important;
+  width: 32px !important;
+  margin: 8px !important;
+  background-image: url('data:image/svg+xml;charset=utf-8,<svg width="32" height="32" viewBox="0 0 32 32" fill="none" xmlns="http://www.w3.org/2000/svg"> <path d="M16 32C24.8366 32 32 24.8366 32 16C32 7.16344 24.8366 0 16 0C7.16344 0 0 7.16344 0 16C0 24.8366 7.16344 32 16 32Z" fill="white"/> <path d="M6.74612 18.796C6.33376 18.0797 6.57889 17.1649 7.29412 16.7508L15.0632 12.2529C15.7808 11.8374 16.6995 12.0835 17.1132 12.8022C17.5256 13.5184 17.2805 14.4333 16.5652 14.8474L8.7962 19.3453C8.07854 19.7608 7.15989 19.5146 6.74612 18.796Z" fill="black"/> <path d="M14.724 12.8022C15.1377 12.0835 16.0564 11.8374 16.774 12.2528L24.5431 16.7508C25.2583 17.1648 25.5034 18.0797 25.0911 18.7959C24.6773 19.5146 23.7587 19.7607 23.041 19.3452L15.272 14.8473C14.5567 14.4332 14.3116 13.5184 14.724 12.8022Z" fill="black"/> </svg>') !important;
+  background-position: center;
+}
+
+.i-amphtml-story-inline-page-attachment-label {
+  display: none !important;
```
Comment (processprocess) [inline extensions/amp-story/1.0/amp-story-open-page-attachment.css:215]: We don't need this anymore since we're dynamically appending it.

File: extensions/amp-story/1.0/amp-story-open-page-attachment.js [hunk 7]
```diff
@@ -65,3 +104,59 @@ export const renderPageAttachmentUI = (pageEl, attachmentEl) => {
...
+
+  const openLabelAttr = attachmentEl.getAttribute('cta-text');
+  if (openLabelAttr) {
+    const textEl = win.document.createElement('span');
+    textEl.classList.add("i-amphtml-story-inline-page-attachment-label");
+    const openLabel = openLabelAttr && openLabelAttr.trim();
+    textEl.textContent = openLabel;
+    setImportantStyles(textEl, {
+      'display': 'block',
+    });
```
Comment (processprocess) [inline extensions/amp-story/1.0/amp-story-open-page-attachment.js:126]: We don't need this anymore since we're dynamically appending it.

File: extensions/amp-story/1.0/amp-story-open-page-attachment.js [hunk 8]
```diff
@@ -65,3 +104,59 @@ export const renderPageAttachmentUI = (pageEl, attachmentEl) => {
...
+
+  if (openImgAttr2) {
+    const ctaImgEl2 = openAttachmentEl.querySelector(
+      '.i-amphtml-story-inline-page-attachment-img-2'
+    );
+    setImportantStyles(ctaImgEl2, {
+      'display': 'block',
+      'background-image': 'url(' + openImgAttr2 + ')'
+    });
+  }
```
Comment (processprocess) [inline extensions/amp-story/1.0/amp-story-open-page-attachment.js:150]: Since these are optional we should build them only if they're  specified.
We should always try to keep our dom tree as small as possible.

File: extensions/amp-story/1.0/amp-story-open-page-attachment.css [hunk 9]
```diff
@@ -166,3 +166,61 @@
...
+.i-amphtml-story-inline-page-attachment-img,
+.i-amphtml-story-inline-page-attachment-img-2 {
+  height: 48px !important;
+  width: 48px !important;
+  background-color: white !important;
+}
+
+.i-amphtml-story-inline-page-attachment-img-2 {
+  display: none !important;
+  margin-left: 1px !important;
```
Comment (processprocess) [inline extensions/amp-story/1.0/amp-story-open-page-attachment.css:203]: If building these dynamically (comment in amp-story-open-page-attachment.js),
these can have the same class  `i-amphtml-story-inline-page-attachment-img` and the second child can be targeted using `.i-amphtml-story-inline-page-attachment-img:nth-child(2)`.
This way the `i-amphtml-story-inline-page-attachment-img-2` class can be removed.

File: extensions/amp-story/1.0/amp-story-open-page-attachment.css [hunk 10]
```diff
@@ -166,3 +166,61 @@
...
   white-space: nowrap !important;
 }
+
+/** Inline attachment specifications */
+.i-amphtml-story-inline-page-attachment-chip {
+  display: flex !important;
+  position: relative !important;
+  border-radius: 24px !important;
+  background-color: rgba(255, 255, 255, 0.5) !important;
+  color: black !important;
```
Comment (processprocess) [inline extensions/amp-story/1.0/amp-story-open-page-attachment.css:176]: Optional structure nit:
It's good practice to use css variables for color. It makes maintenance easier.

It's also helpful with theming. Variables can be overridden depending on theme and it will then automatically update wherever the variable is used.
It might be worth setting this up now.

File: extensions/amp-story/1.0/amp-story-open-page-attachment.js (hunk 2 above)
Comment (processprocess) [reply extensions/amp-story/1.0/amp-story-open-page-attachment.js:53]: Update, it's `aria-label` not, `alt`. My mistake :) 
[role-img docs](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/Role_Img)

File: extensions/amp-story/1.0/amp-story-open-page-attachment.js (hunk 8 above)
Comment (raxsha) [reply extensions/amp-story/1.0/amp-story-open-page-attachment.js:150]: The renderPageAttachmentUI method checks for 'cta-img' bc if there are no images specified, we fallback on the old CTA button design.

File: extensions/amp-story/1.0/amp-story-open-page-attachment.css (hunk 4 above)
Comment (raxsha) [reply extensions/amp-story/1.0/amp-story-open-page-attachment.css:173]: position isn't needed, display is (or else the images display in a single column instead of a row)

File: extensions/amp-story/1.0/amp-story-open-page-attachment.css (hunk 5 above)
Comment (raxsha) [reply extensions/amp-story/1.0/amp-story-open-page-attachment.css:172]: they become a single column instead of a row without it

File: examples/amp-story/attachment.html [hunk 11]
```diff
@@ -51,7 +51,16 @@
...
-        <amp-story-page-attachment data-title="Lorem ipsum" layout="nodisplay">
+        <amp-story-page-attachment
+          data-title="Lorem ipsum"
+          layout="nodisplay"
+          cta-text="Call To Action"
+          cta-img="/extensions/amp-story/img/sweater.jpg"
+          cta-img-alt="sweater"
+          cta-img-2="/extensions/amp-story/img/boot.jpg"
+          cta-img-2-alt="boot"
+          theme="light"
```
Comment (processprocess) [inline examples/amp-story/attachment.html:62]: This can be removed and handled in a separate PR for #32771

File: extensions/amp-story/1.0/amp-story-open-page-attachment.css [hunk 12]
```diff
@@ -166,3 +166,56 @@
...
+  margin-bottom: 12px !important;
+}
+
+.i-amphtml-story-inline-page-attachment-chip:only-child {
+  margin-bottom: 20px !important;
+}
+
+.i-amphtml-story-inline-page-attachment-img,
+.i-amphtml-story-inline-page-attachment-arrow {
+  vertical-align: middle !important;
```
Comment (processprocess) [inline extensions/amp-story/1.0/amp-story-open-page-attachment.css:187]: Might not need this.

File: extensions/amp-story/1.0/amp-story-open-page-attachment.css [hunk 13]
```diff
@@ -166,3 +166,56 @@
...
+  background-size: contain !important;
+  background-repeat: no-repeat !important;
+}
+
+.i-amphtml-story-inline-page-attachment-img {
+  height: 48px !important;
+  width: 48px !important;
+  background-color: white !important;
```
Comment (processprocess) [inline extensions/amp-story/1.0/amp-story-open-page-attachment.css:195]: comment:
This could be a color variable. 
Let's do that in the theming PR.

File: extensions/amp-story/1.0/amp-story-open-page-attachment.css [hunk 14]
```diff
@@ -166,3 +166,56 @@
   text-shadow: 0px 0px 6px rgba(0, 0, 0, 0.36) !important;
   white-space: nowrap !important;
 }
+
+/** For amp-story-inline-page-attachment-v2 experiment elements. */
+.i-amphtml-story-inline-page-attachment-chip {
+  --cta-chip-bg-color: rgba(255, 255, 255, 0.5);
+  background-color: var(--cta-chip-bg-color) !important;
```
Comment (processprocess) [inline extensions/amp-story/1.0/amp-story-open-page-attachment.css:173]: Let's save the variables for the theming PR actually. 
That way we can see it all working together with the theme attribute.

File: extensions/amp-story/1.0/amp-story-open-page-attachment.js [hunk 15]
```diff
@@ -38,13 +40,49 @@ export const buildOpenAttachmentElement = (element) =>
...
 /**
- * Renders the open attachment UI affordance.
+ * @param {!Element} element
+ * @return {!Element}
+ */
+export const buildOpenInlineAttachmentElement = (element) =>
+  htmlFor(element)`
+    <a class="
+        i-amphtml-story-page-open-attachment i-amphtml-story-system-reset"
+        role="button">
```
Comment (processprocess) [inline extensions/amp-story/1.0/amp-story-open-page-attachment.js:50]: Since this has the `role="button"`, the `aria-label` needs to be on this element.
It could be a default string like `aria-label="Open page attachment"`.
That would need to be localized string.

This could then be overridden by the `cta-text` attribute.

File: extensions/amp-story/1.0/amp-story-open-page-attachment.js [hunk 16]
```diff
@@ -38,13 +40,49 @@ export const buildOpenAttachmentElement = (element) =>
...
+      <div class="i-amphtml-story-inline-page-attachment-chip">
+        <div role="img" class="i-amphtml-story-inline-page-attachment-img"></div>
+        <div role="img" aria-label="Upward pointing arrow to indicate scroll up to view page attachment" class="i-amphtml-story-inline-page-attachment-arrow"></div>
```
Comment (processprocess) [inline extensions/amp-story/1.0/amp-story-open-page-attachment.js:53]: The aria-label needs to be on the parent `role=button` element actually. 
My bad, I'm realizing this after testing it on a device. We definitely should test a11y on a device while developing for a11y.

We're removing the label here but general feedback about it:
If it's a UI element, it should describe the action, not the image. This one could be something like "Open attachment" since that's the action it describes. It's also good to keep them simple because they need to be translated for i18n.

File: extensions/amp-story/1.0/amp-story-open-page-attachment.js [hunk 17]
```diff
@@ -65,3 +103,70 @@ export const renderPageAttachmentUI = (pageEl, attachmentEl) => {
...
+    setImportantStyles(ctaImgEl2, {
+      'background-image': 'url(' + openImgAttr2 + ')'
+    });
+    const openImgAttrAlt2 = attachmentEl.getAttribute('cta-img-2-alt');
+    if (openImgAttrAlt2) {
+      ctaImgEl2.setAttribute("aria-label", openImgAttrAlt2);
+    }
+    const ctaChipEl = openAttachmentEl.querySelector(
+      '.i-amphtml-story-inline-page-attachment-chip'
+    );
```
Comment (processprocess) [inline extensions/amp-story/1.0/amp-story-open-page-attachment.js:158]: Might not be using this anymore.

File: extensions/amp-story/1.0/amp-story-open-page-attachment.js [hunk 18]
```diff
@@ -65,3 +103,70 @@ export const renderPageAttachmentUI = (pageEl, attachmentEl) => {
...
+
+  if (openImgAttr2) {
+    const ctaImgEl2 = win.document.createElement('span');
+    ctaImgEl2.classList.add("i-amphtml-story-inline-page-attachment-img");
+    setImportantStyles(ctaImgEl2, {
+      'background-image': 'url(' + openImgAttr2 + ')'
+    });
+    const openImgAttrAlt2 = attachmentEl.getAttribute('cta-img-2-alt');
+    if (openImgAttrAlt2) {
+      ctaImgEl2.setAttribute("aria-label", openImgAttrAlt2);
```
Comment (processprocess) [inline extensions/amp-story/1.0/amp-story-open-page-attachment.js:154]: These actually won't be read if the parent element has an `aria-label` attribute. 
I think these can be considered decorative and therefore we don't need to announce them. 
We can remove this attribute logic. I didn't realize this until testing it on a device, it's definitely a good idea to test a11y on a device while developing for something like this.

File: extensions/amp-story/validator-amp-story.protoascii [hunk 19]
```diff
@@ -844,6 +844,29 @@ tags: {  # <amp-story-page-attachment> with no href
     value: "dark"
     value: "light"
   }
+  attrs: {
```
Comment (honeybadgerdontcare) [inline extensions/amp-story/validator-amp-story.protoascii:847]: Please alphabetize the attributes by their name.

File: extensions/amp-story/validator-amp-story.protoascii [hunk 20]
```diff
@@ -844,6 +844,29 @@ tags: {  # <amp-story-page-attachment> with no href
...
+    name: "cta-text"
+  }
+  attrs: {
```
Comment (honeybadgerdontcare) [inline extensions/amp-story/validator-amp-story.protoascii:850]: This will allow relative URLs for `cta-img` and `cta-img-2` below as that is the [default for validator](https://github.com/ampproject/amphtml/blob/master/validator/validator.proto#L62). Is that desired? Should AMP caches rewrite these URLs to be absolute?

**Review by honeybadgerdontcare — CHANGES_REQUESTED** [review]

File: extensions/amp-story/1.0/amp-story-open-page-attachment.css (hunk 14 above)
Comment (raxsha) [reply extensions/amp-story/1.0/amp-story-open-page-attachment.css:173]: yeah I agree with that

File: extensions/amp-story/1.0/amp-story-open-page-attachment.css (hunk 13 above)
Comment (raxsha) [reply extensions/amp-story/1.0/amp-story-open-page-attachment.css:195]: noted for theming PR - #32771

File: extensions/amp-story/validator-amp-story.protoascii (hunk 20 above)
Comment (honeybadgerdontcare) [reply extensions/amp-story/validator-amp-story.protoascii:840]: See cl/361937561. Are we confident that these attribute names aren't changing: `cta-img` and `cta-img-2`?

Comment (newmuis) [reply extensions/amp-story/validator-amp-story.protoascii:840]: Ok, I'll bite and bikeshed the name, since this is going to be proliferated across codebases.

Let's use `image` instead of `img`, to align with the [interactive component result API](https://github.com/ampproject/amphtml/tree/master/extensions/amp-story-interactive#multi-page-component-apis).  @honeybadgerdontcare do we need to support that in the transformer as well?

Comment (raxsha) [reply extensions/amp-story/validator-amp-story.protoascii:840]: Yes, they'll stay with those names. We considered different approaches but this is what we need

**Comment by raxsha** [comment]
> Firebase demo:
> The screenshots are great but seeing a working demo is ideal, especially for UI updates.
> I can help you set that up also if you want!

Added! You can view at https://amp-story-inline-cta.web.app/examples/amp-story/attachment.html by toggling the feature flag in the console with: `AMP.toggleExperiment('amp-story-inline-page-attachment-v2');`.

File: extensions/amp-story/validator-amp-story.protoascii (hunk 20 above)
Comment (raxsha) [reply extensions/amp-story/validator-amp-story.protoascii:840]: changed to `cta-image` and `cta-image-2`

Comment (honeybadgerdontcare) [reply extensions/amp-story/validator-amp-story.protoascii:840]: Yes, we'll need to support that explicitly. I'll update the cl.

These are the attributes regardless of which tag they are on that get made absolute:
  * background
  * href
  * xlink:href
  * poster
  * src
  * srcset

File: extensions/amp-story/validator-amp-story.protoascii [hunk 21]
```diff
@@ -834,6 +834,23 @@ tags: {  # <amp-story-page-attachment> with no href
   mandatory_ancestor: "AMP-STORY-PAGE"
   descendant_tag_list: "amp-story-page-attachment-allowed-descendants"
   mandatory_last_child: true
+  attrs: {
+    name: "cta-text"
```
[the remark this answers is not in the document]
Comment (honeybadgerdontcare) [reply extensions/amp-story/validator-amp-story.protoascii:838]: Just fyi, all attributes starting with `data-` are valid. I could put `data-foo` on any AMP tag and it would still be valid.

[the remark this answers is not in the document]
Comment (processprocess) [reply extensions/amp-story/validator-amp-story.protoascii:838]: Got it! Thank you for clarifying :)

File: extensions/amp-story/1.0/amp-story-open-page-attachment.js [hunk 22]
```diff
@@ -38,13 +40,49 @@ export const buildOpenAttachmentElement = (element) =>
...
 /**
- * Renders the open attachment UI affordance.
+ * @param {!Element} element
+ * @return {!Element}
+ */
+export const buildOpenInlineAttachmentElement = (element) =>
+  htmlFor(element)`
+    <a class="
+        i-amphtml-story-page-open-attachment i-amphtml-story-system-reset"
+        role="button" aria-label="Open page attachment">
```
[the remark this answers is not in the document]
Comment (raxsha) [reply extensions/amp-story/1.0/amp-story-open-page-attachment.js]: I used the existing `AMP_STORY_PAGE_ATTACHMENT_OPEN_LABEL` localized string

File: extensions/amp-story/1.0/amp-story-open-page-attachment.js (hunk 15 above)
Comment (newmuis) [reply extensions/amp-story/1.0/amp-story-open-page-attachment.js:50]: We should be careful here; "page attachment" is not a term that we expose to users, so they (likely?) won't know what this means.

We normally (thus far) have erred on the side of "swipe up" or "read more" or similar.

File: extensions/amp-story/validator-amp-story.protoascii (hunk 21 above)
[the remark this answers is not in the document]
Comment (raxsha) [reply extensions/amp-story/validator-amp-story.protoascii]: If it has a `data-` prefix it doesn't need to be validated, IIRC, so I'll remove it from the validator here. I 

ind of think it should have always been without the `data-` prefix but you're right, this should continue to be `data-cta-text` to be consistent with the docs. I think the new `cta-image` attributes should not have that prefix though bc they should be validated.

[the remark this answers is not in the document]
Comment (processprocess) [reply extensions/amp-story/validator-amp-story.protoascii]: > cta-image attributes should not have that prefix though bc they should be validated.

I agree but I'll default to @gmajoulet and @newmuis

File: extensions/amp-story/validator-amp-story.protoascii (hunk 20 above)
Comment (raxsha) [reply extensions/amp-story/validator-amp-story.protoascii]: They'll be `data-cta-image` and `data-cta-image-2`, actually

Comment (honeybadgerdontcare) [reply extensions/amp-story/validator-amp-story.protoascii]: Updated transformers to reflect `data-cta-image` and `data-cta-image-2`.
Updated the protoascii here adding `allow_relative: false` until the transformers are in prod.

**Review by processprocess — APPROVED** [review]

**Review by honeybadgerdontcare — APPROVED** [review]
validation looks good. can remove allow_relative once transformers are live in cache.

File: extensions/amp-story/validator-amp-story.protoascii (hunk 20 above)
Comment (raxsha) [reply extensions/amp-story/validator-amp-story.protoascii]: final update here: we used `cta-image` and `cta-image-2`. Thank you so much David!

File: tools/experiments/experiments-config.js [hunk 23]
```diff
@@ -220,6 +220,11 @@ export const EXPERIMENTS = [
     name: 'Client-side granular consent',
     spec: 'https://github.com/ampproject/amphtml/issues/31607',
   },
+  {
+    id: 'amp-story-inline-page-attachment-v2',
```
Comment (gmajoulet) [inline tools/experiments/experiments-config.js:224]: We should make the experiment global to all changes to page attachments, so we can launch everything at once when ready.

File: extensions/amp-story/validator-amp-story.protoascii [hunk 24]
```diff
@@ -834,6 +834,22 @@ tags: {  # <amp-story-page-attachment> with no href
   mandatory_ancestor: "AMP-STORY-PAGE"
   descendant_tag_list: "amp-story-page-attachment-allowed-descendants"
   mandatory_last_child: true
+  attrs: {
```
Comment (gmajoulet) [inline extensions/amp-story/validator-amp-story.protoascii:837]: Can you add validation tests?

File: extensions/amp-story/1.0/amp-story-open-page-attachment.js [hunk 25]
```diff
@@ -38,13 +40,49 @@ export const buildOpenAttachmentElement = (element) =>
...
+  const openImgAttr = attachmentEl.getAttribute('cta-image');
+  const attachmentHref = attachmentEl.getAttribute('href');
+  if (
+    isInlinePageAttachmentV2ExperimentOn(win) &&
+    !attachmentHref &&
+    openImgAttr
+  ) {
+    return renderInlinePageAttachmentUI(win, pageEl, attachmentEl);
+  } else {
+    return renderDefaultPageAttachmentUI(pageEl, attachmentEl);
```
Comment (gmajoulet) [inline extensions/amp-story/1.0/amp-story-open-page-attachment.js:74]: If we're going to maintain two templates (one with images, one without), let's update the `renderInlinePageAttachmentUI` method name with something that reflects what it builds: the attachment UI that has images. Both are for the inline attachment so the names are confusing.

File: extensions/amp-story/1.0/amp-story-open-page-attachment.js [hunk 26]
```diff
@@ -65,3 +103,57 @@ export const renderPageAttachmentUI = (pageEl, attachmentEl) => {
...
+
+  const openLabelAttr = attachmentEl.getAttribute('data-cta-text');
+  if (openLabelAttr) {
+    const textEl = win.document.createElement('span');
+    textEl.classList.add('i-amphtml-story-inline-page-attachment-label');
+    const openLabel = openLabelAttr && openLabelAttr.trim();
+    textEl.textContent = openLabel;
+    openAttachmentEl.appendChild(textEl);
+    // Override descriptive text on page-attachment button.
+    openAttachmentEl.setAttribute('aria-label', openLabel);
```
Comment (gmajoulet) [inline extensions/amp-story/1.0/amp-story-open-page-attachment.js:125]: This is a regression, we're losing i18n here.
It should be the same as above

```
const openLabelAttr = attachmentEl.getAttribute('data-cta-text');
  const openLabel =
    (openLabelAttr && openLabelAttr.trim()) ||
    getLocalizationService(pageEl).getLocalizedString(
      LocalizedStringId.AMP_STORY_PAGE_ATTACHMENT_OPEN_LABEL
    );
```

File: extensions/amp-story/1.0/amp-story-open-page-attachment.js (hunk 25 above)
Comment (gmajoulet) [reply extensions/amp-story/1.0/amp-story-open-page-attachment.js:74]: We're duplicating a lot of code that we won't be able to remove when launching the experiment. Is it possible to have only one template for the new UI, so we can remove the old one when the experiment is launched?

File: extensions/amp-story/1.0/amp-story-open-page-attachment.js [hunk 27]
```diff
@@ -38,13 +40,49 @@ export const buildOpenAttachmentElement = (element) =>
...
 /**
- * Renders the open attachment UI affordance.
+ * @param {!Element} element
+ * @return {!Element}
+ */
+export const buildOpenInlineAttachmentElement = (element) =>
+  htmlFor(element)`
+    <a class="
+        i-amphtml-story-page-open-attachment i-amphtml-story-system-reset"
+        role="button" aria-label="Swipe up">
```
Comment (gmajoulet) [inline extensions/amp-story/1.0/amp-story-open-page-attachment.js:50]: Please remove the untranslated aria-label, it shouldn't be used, we add it programmatically using the i18n service.

Comment (processprocess) [reply extensions/amp-story/1.0/amp-story-open-page-attachment.js:50]: Is there ever a case were the service wouldn't load and this would be helpful as a fallback?

Comment (gmajoulet) [reply extensions/amp-story/1.0/amp-story-open-page-attachment.js:50]: I don't think so, but tbh I'm mostly worried about duplicating the strings and/or forgetting to i18n like we did in this PR.

File: extensions/amp-story/1.0/amp-story-open-page-attachment.js (hunk 26 above)
Comment (raxsha) [reply extensions/amp-story/1.0/amp-story-open-page-attachment.js]: Something new in the new UI is that there is an option for no text with the button, whereas previously we would default to this `AMP_STORY_PAGE_ATTACHMENT_OPEN_LABEL` text if none was provided. So because of that, the way I've changed this now is that the `aria-label` is now either:

1) the provided text from the publisher, matching the label on the CTA button.
2) the default `AMP_STORY_PAGE_ATTACHMENT_OPEN_LABEL` text, which will be the accessibility text when there's no visible label provided.

File: extensions/amp-story/1.0/amp-story-open-page-attachment.js (hunk 25 above)
Comment (raxsha) [reply extensions/amp-story/1.0/amp-story-open-page-attachment.js:70]: Updated it to `renderPageAttachmentUiWithImages`

Comment (raxsha) [reply extensions/amp-story/1.0/amp-story-open-page-attachment.js:70]: Unsure about that second part-- the way I see it is that we currently have:
1. old CTA UI (for outlinks & when no images are provided for inline)
2. inline CTA UI with images

And soon we'll have:
1. old CTA UI (when no images are provided for inline)
2. inline CTA UI with images
3. new outlink UI

I don't believe we'll ever want to remove the old UI completely, since it will still be used when no images are provided in-line, at least from what I understand from Hong's slides.

When the experiment is launched fully, all we'll be doing is changing the conditional logic in `renderPageAttachmentUI` to render one of these 3 UIs.

File: extensions/amp-story/1.0/amp-story-open-page-attachment.css [hunk 28]
```diff
@@ -166,3 +166,54 @@
...
+}
+
+.i-amphtml-story-inline-page-attachment-img {
+  height: 48px !important;
+  width: 48px !important;
+  background-color: white !important;
+}
+
+.i-amphtml-story-inline-page-attachment-img:nth-child(2) {
+  margin-left: 1px !important;
```
[the remark this answers is not in the document]
Comment (processprocess) [reply extensions/amp-story/1.0/amp-story-open-page-attachment.css:197]: If we use a flex layout wouldn't the margin need to be an element?
I wonder if overriding in RTL would be sufficient:
```
[dir=rtl] .i-amphtml-story-inline-page-attachment-img:nth-child(2) {
   margin-left: 0 !important;
   margin-right: 1px !important;
}
```

[the remark this answers is not in the document]
Comment (raxsha) [reply extensions/amp-story/1.0/amp-story-open-page-attachment.css]: dir=rtl doesn't work when direction is set outside the shadow DOM
But `margin-inline-start: 1px !important;` works-- it sets a margin on the side of the element that it determines is the "start" so it works well in this case.

File: extensions/amp-story/1.0/amp-story-open-page-attachment.js [hunk 29]
```diff
@@ -38,13 +40,45 @@ export const buildOpenAttachmentElement = (element) =>
...
+ * @param {!Window} win
+ * @param {!Element} pageEl
+ * @param {!Element} attachmentEl
+ * @return {!Element}
+ */
+export const renderPageAttachmentUI = (win, pageEl, attachmentEl) => {
+  const openImgAttr = attachmentEl.getAttribute('cta-image');
+  const attachmentHref = attachmentEl.getAttribute('href');
+  if (isPageAttachmentUiV2ExperimentOn(win) && !attachmentHref && openImgAttr) {
+    return renderPageAttachmentUiWithImages(win, pageEl, attachmentEl);
```
Comment (processprocess) [inline extensions/amp-story/1.0/amp-story-open-page-attachment.js:68]: It would be good to test that this codepath runs and correct UI renders.

We could write two tests for this: 
- Renders new UI if the experiment is on and one image is passed
- Renders new UI if the experiment is on and two images is passed

File: extensions/amp-story/1.0/amp-story-open-page-attachment.js (hunk 25 above)
Comment (processprocess) [reply extensions/amp-story/1.0/amp-story-open-page-attachment.js]: Gabriel does make a good point. It could work for this to be one template.
The template could default as no-image UI and append image nodes if they're specified.

I think it would be more lean after launch but a little more code up-front since we would need to build the no-image UI into the experiment.

File: extensions/amp-story/1.0/amp-story-open-page-attachment.js (hunk 26 above)
Comment (processprocess) [reply extensions/amp-story/1.0/amp-story-open-page-attachment.js]: We could not override the i18n string but that assumes that the publisher's text is decorative. 

I wonder if it would work to prepend the publisher specified string to the i18n label.

File: extensions/amp-story/1.0/amp-story-open-page-attachment.js (hunk 25 above)
Comment (processprocess) [reply extensions/amp-story/1.0/amp-story-open-page-attachment.js]: I think Gabriel's suggestion is good since it reduces code and branching.

Here is an example commit of how this could be one template if you would like to reference it:
https://github.com/ampproject/amphtml/pull/33235/commits/642d29550c006a1d29223fa2d9e3bceec2eb9535

Also added some code comments since our function is doing several things.

File: extensions/amp-story/1.0/amp-story-open-page-attachment.js (hunk 29 above)
Comment (raxsha) [reply extensions/amp-story/1.0/amp-story-open-page-attachment.js:68]: done!

File: extensions/amp-story/1.0/amp-story-open-page-attachment.js (hunk 26 above)
Comment (raxsha) [reply extensions/amp-story/1.0/amp-story-open-page-attachment.js]: Ok so the solution we've gone with now is using the default i18n "Swipe up" text for the component and the aria-label unless the publisher specifies text or specifies 'none'.

File: extensions/amp-story/1.0/amp-story-open-page-attachment.js (hunk 25 above)
Comment (raxsha) [reply extensions/amp-story/1.0/amp-story-open-page-attachment.js]: Philip and I investigated this and the way to collapse the templates when there's going to be 3 CTA UIs would be overly complicated and may cause regressions with the existing animations or other regressions. So I think we should keep the structure we have for now, because changing it would add complexity to readability without much benefit.

**Review by gmajoulet — APPROVED** [review]
